File: src/main/java/com/now/nowbot/model/json/Match.java
package com.now.nowbot.model.json;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.now.nowbot.model.enums.OsuMod;
import org.jetbrains.annotations.NotNull;
import org.springframework.lang.NonNull;
import org.springframework.lang.Nullable;
import org.springframework.util.CollectionUtils;

import java.time.OffsetDateTime;
import java.util.*;

@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
@JsonIgnoreProperties(ignoreUnknown = true)
public class Match implements Cloneable {
    @JsonProperty("match")
    MatchStat matchStat;

    @JsonProperty("events")
    List<MatchEvent> events;

    @JsonProperty("users")
    List<MicroUser> players;

    @JsonProperty("first_event_id")
    Long firstEventID;

    @JsonProperty("latest_event_id")
    Long latestEventID;

    @JsonProperty("current_game_id")
    Long currentGameID;

    public MatchStat getMatchStat() {
        return matchStat;
    }

    public void setMatchStat(MatchStat matchStat) {
        this.matchStat = matchStat;
    }

    public List<MatchEvent> getEvents() {
        return events;
    }

    public void setEvents(List<MatchEvent> events) {
        this.events = events;
    }

    public List<MicroUser> getPlayers() {
        return players;
    }

    public void setPlayers(List<MicroUser> players) {
        this.players = players;
    }

    public Long getFirstEventID() {
        return firstEventID;
    }

    public void setFirstEventID(Long firstEventID) {
        this.firstEventID = firstEventID;
    }

    public Long getLatestEventID() {
        return latestEventID;
    }

    public void setLatestEventID(Long latestEventID) {
        this.latestEventID = latestEventID;
    }

    public Long getCurrentGameID() {
        return currentGameID;
    }

    public void setCurrentGameID(Long currentGameID) {
        this.currentGameID = currentGameID;
    }

    @Override
    public Match clone() {
        try {
            return (Match) super.clone();
        } catch (CloneNotSupportedException e) {
            throw new AssertionError();
        }
    }

    @Override
    public String toString() {
        return STR."Match{matchStat=\{matchStat}, events=\{events}, players=\{players}, firstEventId=\{firstEventID}, latestEventId=\{latestEventID}, currentGameId=\{currentGameID}\{'}'}";
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public static class MatchStat {
        @JsonProperty("id")
        Long matchID;

        @JsonProperty("start_time")
        OffsetDateTime startTime;

        @JsonProperty("end_time")
        OffsetDateTime endTime;

        String name;

        public Long getMatchID() {
            return matchID;
        }

        public void setMatchID(Long matchID) {
            this.matchID = matchID;
        }

        public OffsetDateTime getStartTime() {
            return startTime;
        }

        public void setStartTime(OffsetDateTime startTime) {
            this.startTime = startTime;
        }

        public OffsetDateTime getEndTime() {
            return endTime;
        }

        public void setEndTime(OffsetDateTime endTime) {
            this.endTime = endTime;
        }

        public String getName() {
            return name;
        }

        public void setName(String name) {
            this.name = name;
        }

        @Override
        public String toString() {
            return STR."MatchStat{matchID=\{matchID}, startTime=\{startTime}, endTime=\{endTime}, name='\{name}\{'\''}\{'}'}";
        }
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public static class MatchEvent implements Comparable<MatchEvent> {

        public record Detail(String type, String text) {}

        @JsonProperty("id")
        Long eventID;

        @JsonProperty("detail")
        Detail detail;

        OffsetDateTime timestamp;

        @JsonProperty("user_id")
        Long userID;

        @JsonProperty("game")
        MatchRound round;

        public Long getEventID() {
            return eventID;
        }

        public void setEventID(Long eventID) {
            this.eventID = eventID;
        }

        public Detail getDetail() {
            return detail;
        }

        public void setDetail(Detail detail) {
            this.detail = detail;
        }

        public OffsetDateTime getTimestamp() {
            return timestamp;
        }

        public void setTimestamp(OffsetDateTime timestamp) {
            this.timestamp = timestamp;
        }

        public Long getUserID() {
            return userID;
        }

        public void setUserID(Long userID) {
            this.userID = userID;
        }

        public MatchRound getRound() {
            return round;
        }

        public void setRound(MatchRound round) {
            this.round = round;
        }

        @Override
        public String toString() {
            return STR."MatchEvent{eventID=\{eventID}, detail=\{detail}, timestamp=\{timestamp}, userID=\{userID}, round=\{round}\{'}'}";
        }

        @Override
        public int compareTo(@NotNull MatchEvent o) {
            return Math.toIntExact(o.eventID - eventID);
        }
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public static class MatchRound {
        @JsonProperty("id")
        Integer roundID;

        @JsonProperty("beatmap_id")
        Long beatMapID;

        @JsonProperty("start_time")
        OffsetDateTime startTime;

        @JsonProperty("end_time")
        OffsetDateTime endTime;

        String mode;

        @Nullable
        @JsonProperty("mod_int")
        Integer modInt;

        @JsonProperty("scoring_type")
        String scoringType;

        @JsonProperty("team_type")
        String teamType;

        List<String> mods;

        @Nullable
        @JsonProperty("beatmap")
        BeatMap beatMap;

        @JsonProperty("scores")
        List<MatchScore> scores;

        public Long getBeatMapID() {
            return beatMapID;
        }

        public void setBeatMapID(Long beatMapID) {
            this.beatMapID = beatMapID;
        }

        public Integer getRoundID() {
            return roundID;
        }

        public void setRoundID(Integer roundID) {
            this.roundID = roundID;
        }

        public OffsetDateTime getStartTime() {
            return startTime;
        }

        public void setStartTime(OffsetDateTime startTime) {
            this.startTime = startTime;
        }

        public OffsetDateTime getEndTime() {
            return endTime;
        }

        public void setEndTime(OffsetDateTime endTime) {
            this.endTime = endTime;
        }

        public String getMode() {
            return mode;
        }

        public void setMode(String mode) {
            this.mode = mode;
        }

        public Integer getModInt() {
            return Optional.ofNullable(modInt).orElse(OsuMod.getModsValueFromAbbrList(this.mods));
        }

        public void setModInt(@Nullable Integer modInt) {
            this.modInt = modInt;
        }

        public String getScoringType() {
            return scoringType;
        }

        public void setScoringType(String scoringType) {
            this.scoringType = scoringType;
        }

        public String getTeamType() {
            return teamType;
        }

        public void setTeamType(String teamType) {
            this.teamType = teamType;
        }

        public List<String> getMods() {
            return mods;
        }

        public void setMods(List<String> mods) {
            this.mods = mods;
        }

        @Nullable
        public BeatMap getBeatMap() {
            return beatMap;
        }

        public void setBeatMap(@Nullable BeatMap beatMap) {
            this.beatMap = beatMap;
        }

        public List<MatchScore> getScores() {
            return scores;
        }

        public void setScores(List<MatchScore> scores) {
            this.scores = scores;
        }

        public record TeamScore(Long total, Long red, Long blue) {}

        public TeamScore getTeamScore() {
            long total = 0L, red = 0L, blue = 0L;

            if (Objects.equals(teamType, "team-vs")) {
                for (MatchScore s : scores) {

                    switch (s.getPlayerStat().team()) {
                        case "red" -> red += s.getScore();
                        case "blue" -> blue += s.getScore();
                    }
                }

                total = red + blue;

            } else {
                for (MatchScore s : scores) {
                    total += s.getScore();
                }
            }

            return new TeamScore(total, red, blue);
        }

        @Nullable
        public String getWinningTeam() {
            if (! Objects.equals(this.getTeamType(), "team-vs")) return "none";
            var ts = getTeamScore();

            if (ts.red() > ts.blue()) return "red";
            else if (ts.red() < ts.blue()) return "blue";
            else return null; //平局
        }

        @NonNull
        public Long getWinningTeamScore() {
            if (Objects.equals(teamType, "team-vs")) {
                var ts = getTeamScore();
                return Math.max(ts.red(), ts.blue());
            } else {
                return this.getScores().stream().mapToLong(MatchScore::getScore).reduce(Long::max).orElse(0L);
            }
        }

        @Override
        public String toString() {
            return STR."MatchRound{beatMapID=\{beatMapID}, roundID=\{roundID}, startTime=\{startTime}, endTime=\{endTime}, mode='\{mode}\{'\''}, modInt=\{modInt}, scoringType='\{scoringType}\{'\''}, teamType='\{teamType}\{'\''}, mods=\{mods}, beatMap=\{beatMap}, scores=\{scores}\{'}'}";
        }
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public static class MatchScore {
        public record PlayerStat(Integer slot, String team, Boolean pass) {}

        Double accuracy;

        @JsonProperty("best_id")
        Long bestID;

        OffsetDateTime timestamp;

        @JsonProperty("id")
        Long scoreID;

        @JsonProperty("max_combo")
        Integer maxCombo;

        String mode;

        @JsonProperty("mode_int")
        Integer modeInt;

        String[] mods;
        Boolean passed;
        Boolean perfect;
        Double pp;
        String rank;
        Boolean replay;
        Integer score;
        Statistics statistics;
        String type;

        @JsonProperty("user_id")
        Long userID;

        @JsonProperty("match")
        PlayerStat playerStat;

        // 自设
        Integer ranking;

        // 自设
        MicroUser user;

        public Double getAccuracy() {
            return accuracy;
        }

        public void setAccuracy(Double accuracy) {
            this.accuracy = accuracy;
        }

        public Long getBestID() {
            return bestID;
        }

        public void setBestID(Long bestID) {
            this.bestID = bestID;
        }

        public OffsetDateTime getTimestamp() {
            return timestamp;
        }

        public void setTimestamp(OffsetDateTime timestamp) {
            this.timestamp = timestamp;
        }

        public Long getScoreID() {
            return scoreID;
        }

        public void setScoreID(Long scoreID) {
            this.scoreID = scoreID;
        }

        public Integer getMaxCombo() {
            return maxCombo;
        }

        public void setMaxCombo(Integer maxCombo) {
            this.maxCombo = maxCombo;
        }

        public String getMode() {
            return mode;
        }

        public void setMode(String mode) {
            this.mode = mode;
        }

        public Integer getModeInt() {
            return modeInt;
        }

        public void setModeInt(Integer modeInt) {
            this.modeInt = modeInt;
        }

        public String[] getMods() {
            return mods;
        }

        public void setMods(String[] mods) {
            this.mods = mods;
        }

        public Boolean getPassed() {
            return passed;
        }

        public void setPassed(Boolean passed) {
            this.passed = passed;
        }

        public Boolean getPerfect() {
            return perfect;
        }

        public void setPerfect(Boolean perfect) {
            this.perfect = perfect;
        }

        public Double getPp() {
            return pp;
        }

        public void setPp(Double pp) {
            this.pp = pp;
        }

        public String getRank() {
            return rank;
        }

        public void setRank(String rank) {
            this.rank = rank;
        }

        public Boolean getReplay() {
            return replay;
        }

        public void setReplay(Boolean replay) {
            this.replay = replay;
        }

        public Integer getScore() {
            return score;
        }

        public void setScore(Integer score) {
            this.score = score;
        }

        public Statistics getStatistics() {
            return statistics;
        }

        public void setStatistics(Statistics statistics) {
            this.statistics = statistics;
        }

        public String getType() {
            return type;
        }

        public void setType(String type) {
            this.type = type;
        }

        public Long getUserID() {
            return userID;
        }

        public void setUserID(Long userID) {
            this.userID = userID;
        }

        public MatchScore.PlayerStat getPlayerStat() {
            return playerStat;
        }

        public void setPlayerStat(MatchScore.PlayerStat playerStat) {
            this.playerStat = playerStat;
        }

        public Integer getRanking() {
            return ranking;
        }

        public void setRanking(Integer ranking) {
            this.ranking = ranking;
        }

        public MicroUser getUser() {
            return user;
        }

        public void setUser(MicroUser user) {
            this.user = user;
        }

        @Override
        public String toString() {
            return STR."MatchScore{accuracy=\{accuracy}, bestID=\{bestID}, timestamp=\{timestamp}, scoreID=\{scoreID}, maxCombo=\{maxCombo}, mode='\{mode}\{'\''}, modeInt=\{modeInt}, mods=\{Arrays.toString(mods)}, passed=\{passed}, perfect=\{perfect}, pp=\{pp}, rank='\{rank}\{'\''}, replay=\{replay}, score=\{score}, statistics=\{statistics}, type='\{type}\{'\''}, userID=\{userID}, playerStat=\{playerStat}, ranking=\{ranking}, user=\{user}\{'}'}";
        }
    }

    public boolean isMatchEnd() {
        return this.getMatchStat().getEndTime() != null;
    }

    public void parseNextData(Match m) {
        if (CollectionUtils.isEmpty(m.getEvents())) return;
        // 合并事件
        if (events.getFirst().getEventID() > m.getEvents().getLast().getEventID()) {
            // 新事件在前
            this.events.addAll(0, m.getEvents());
        } else if (events.getLast().getEventID() < m.getEvents().getFirst().getEventID()){
            // 新事件在后
            this.events.addAll(m.getEvents());
        } else if (events.getFirst().getEventID() > m.getEvents().getFirst().getEventID()) {
            // 在中间
            events.removeIf(e -> e.getEventID() <= m.getEvents().getLast().getEventID());
            this.events.addAll(0, m.getEvents());
        } else if (events.getLast().getEventID() < m.getEvents().getLast().getEventID()){
            // 在中间
            events.removeIf(e -> e.getEventID() >= m.getEvents().getFirst().getEventID());
            this.events.addAll(m.getEvents());
        }

        if (CollectionUtils.isEmpty(this.players)) {
            this.players = m.getPlayers();
        } else {
            var p = new ArrayList<MicroUser>(this.players.size());

            p.addAll(this.players);
            p.addAll(m.getPlayers());

            this.players = p.stream().distinct().toList();
        }

        //更新状态
        this.matchStat = m.getMatchStat();
        this.latestEventID = m.getLatestEventID();
        this.firstEventID = m.getFirstEventID();
    }

    public static Match mergeMultipleMatches(List<Match> matches) {
        var m0 = new Match();

        for (var m : matches) {
            m0 = merge2Matches(m0, m);
        }

        return m0;
    }

    public static Match merge2Matches(Match m1, Match m2) {
        var m = m1.clone();
        m.getEvents().addAll(m2.getEvents());
        m.setFirstEventID(Math.min(m1.getFirstEventID(), m2.getFirstEventID()));
        m.setLatestEventID(Math.max(m1.getLatestEventID(), m2.getLatestEventID()));

        m.getPlayers().addAll(m2.getPlayers());
        m.setPlayers(m.getPlayers().stream().distinct().toList());

        m.getMatchStat().setStartTime(
                m1.getMatchStat().getStartTime().isBefore(
                        m2.getMatchStat().getStartTime()) ? m1.getMatchStat().getStartTime() : m2.getMatchStat().getStartTime()
        );
        m.getMatchStat().setEndTime(
                m1.getMatchStat().getEndTime().isAfter(
                        m2.getMatchStat().getEndTime()) ? m1.getMatchStat().getEndTime() : m2.getMatchStat().getEndTime()
        );

        m.getMatchStat().setMatchID(m1.getMatchStat().getMatchID());
        m.getMatchStat().setName(m1.getMatchStat().getName());

        return m;
    }
}

File: src/main/java/com/now/nowbot/model/multiplayer/Match.kt
package com.now.nowbot.model.multiplayer

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonProperty
import com.now.nowbot.model.enums.OsuMode
import com.now.nowbot.model.osu.Beatmap
import com.now.nowbot.model.osu.LazerScore
import com.now.nowbot.model.osu.MicroUser
import java.time.OffsetDateTime
import kotlin.math.max

data class Match(
    @JsonProperty("match") var statistics: MatchStat,
    val events: MutableList<MatchEvent>,
    @JsonProperty("users")
    val players: MutableList<MicroUser>,
    @JsonProperty("first_event_id") var firstEventID: Long,
    @JsonProperty("latest_event_id") var latestEventID: Long,
) {
    @get:JsonProperty("is_match_end")
    val isMatchEnd: Boolean
        get() = statistics.endTime != null

    @get:JsonProperty("current_game_id")
    val currentGameID: Long? by lazy {
        val e = events.lastOrNull { it.round != null } ?: return@lazy null
        if (e.round?.endTime == null) return@lazy e.round?.roundID
        return@lazy null
    }

    val name by statistics::name

    val id by statistics::matchID

    val startTime by statistics::startTime

    val endTime by statistics::endTime

    data class MatchStat(
        @JsonProperty("id") val matchID: Long,
        @JsonProperty("start_time") val startTime: OffsetDateTime,
        @JsonProperty("end_time") val endTime: OffsetDateTime?,
        var name: String,
    )

    data class MatchEvent(
        @JsonProperty("id") val eventID: Long,
        @JsonProperty("detail") val detail: MatchEventDetail,
        @JsonProperty("timestamp") val timestamp: OffsetDateTime,
        @JsonProperty("user_id") val userID: Long?,
        @JsonProperty("game") val round: MatchRound?,
    ) {
        val type: EventType
            get() = EventType.getType(detail.type)

        val text: String
            get() = detail.text ?: ""
    }

    data class MatchEventDetail(
        @JsonProperty("type") val type: String,
        @JsonProperty("text") val text: String?,
    )

    data class MatchRound(
        @JsonProperty("id") val roundID: Long,
        @JsonProperty("beatmap") var beatmap: Beatmap?,
        @JsonProperty("beatmap_id") val beatmapID: Long,
        @JsonProperty("start_time") val startTime: OffsetDateTime,
        @JsonProperty("end_time") val endTime: OffsetDateTime?,
        val modeInt: Int,
        val mods: List<String>,
        var scores: List<LazerScore>,
        val teamType: String,
        val scoringType: String,
    ) {
        val mode: OsuMode
            get() = OsuMode.getMode(modeInt)

        @get:JsonProperty("is_team_vs")
        val isTeamVS: Boolean
            get() = teamType == "team-vs" || teamType == "tag-team-vs"

        @get:JsonProperty("red_team_score")
        val redTeamScore: Long
            get() = getTeamScore("red")

        @get:JsonProperty("blue_team_score")
        val blueTeamScore: Long
            get() = getTeamScore("blue")

        @get:JsonProperty("total_team_score")
        val totalTeamScore: Long
            get() = getTeamScore(null)

        @get:JsonIgnore
        val maxScore: Long
            get() = scores.maxOf { it.score }.toLong()

        @get:JsonProperty("winning_team")
        val winningTeam: String?
            get() =
                if (isTeamVS) {
                    val red = getTeamScore("red")
                    val blue = getTeamScore("blue")

                    if (red > blue) "red"
                    else if (red < blue) "blue"
                    else null
                } else {
                    "none"
                }

        @get:JsonProperty("winning_team_score")
        val winningTeamScore: Long
            get() =
                if (isTeamVS) {
                    max(redTeamScore, blueTeamScore)
                } else {
                    totalTeamScore
                }

        private fun getTeamScore(teamType: String? = null): Long {
            return when(teamType) {
                "red" -> scores
                    .filter { it.playerStat?.team == "red" }
                    .sumOf { it.score }
                "blue" -> scores
                    .filter { it.playerStat?.team == "blue" }
                    .sumOf { it.score }
                else -> scores.sumOf { it.score }
            }
        }
    }

    // 2025 05 30 这个类被归并了
    /*
    data class MatchScore(
        @JsonProperty("match") val playerStat: MatchScorePlayerStat,
        @JsonProperty("best_id") val bestID: Long,
        @JsonProperty("user_id") val userID: Long,
        @JsonProperty("id") val scoreID: Long,
        @JsonProperty("accuracy") val accuracy: Double?,
        @JsonProperty("max_combo") val maxCombo: Int,
        @JsonProperty("mode") val mode: String,
        @JsonProperty("mode_int") val modeInt: Int,
        val mods: List<LazerMod>,
        val passed: Boolean,
        val perfect: Boolean,
        val replay: Boolean,
        val pp: Double,
        val rank: String,
        var score: Int,
        val statistics: Statistics,
        val type: String,
    ) {
        // 自己设
        var user: MicroUser? = null
        var ranking: Int? = null
    }

     */

    enum class EventType(val value: String) {
        PlayerJoined("player-joined"),
        PlayerKicked("player-kicked"),
        PlayerLeft("player-left"),
        Other("other"),
        HostChanged("host-changed"),
        MatchDisbanded("match-disbanded"),
        MatchCreated("match-created");

        companion object {
            fun getType(value: String) = when (value) {
                PlayerJoined.value -> PlayerJoined
                PlayerKicked.value -> PlayerKicked
                PlayerLeft.value -> PlayerLeft
                HostChanged.value -> HostChanged
                MatchDisbanded.value -> MatchDisbanded
                MatchCreated.value -> MatchCreated
                else -> Other
            }
        }
    }

    /*
    fun toMatch(): LegacyMatch {
        val match = LegacyMatch()
        match.players = this.players
        match.currentGameID = this.currentGameID
        match.firstEventID = this.firstEventID
        match.latestEventID = this.latestEventID

        val t = LegacyMatch.MatchStat()
        t.matchID = this.id
        t.name = this.name
        t.startTime = this.startTime
        t.endTime = this.endTime
        match.matchStat = t

        val es = mutableListOf<LegacyMatch.MatchEvent>()

        for (me in this.events) {
            val e = LegacyMatch.MatchEvent()

            val r = LegacyMatch.MatchRound()

            if (me.round != null) {
                val g = me.round

                val ss = mutableListOf<LegacyMatch.MatchScore>()

                for (ms in g.scores) {
                    val s = LegacyMatch.MatchScore()

                    val mt = ms.playerStat
                    val pt = LegacyMatch.MatchScore.PlayerStat(mt.slot, mt.team, mt.pass)
                    s.playerStat = pt

                    s.pp = ms.pp
                    s.score = ms.score
                    s.rank = ms.rank
                    s.bestID = ms.bestID
                    s.userID = ms.userID
                    s.scoreID = ms.scoreID
                    s.passed = ms.passed
                    s.accuracy = ms.accuracy
                    s.maxCombo = ms.maxCombo
                    s.mode = ms.mode
                    s.modeInt = ms.modeInt
                    s.mods = ms.mods.toTypedArray()
                    s.perfect = ms.perfect // s.ranking = null
                    s.rank = ms.rank
                    s.replay = ms.replay
                    s.statistics = ms.statistics // s.timestamp = ms.time
                    s.type = ms.type

                    ss.add(s)
                }
                r.scores = ss
                r.mods = g.mods
                r.mode = g.mode.shortname
                r.startTime = g.startTime
                r.endTime = g.endTime
                r.beatmap = g.beatmap
                r.beatmapID = g.beatmapID
                r.modInt = OsuMod.getModsValueFromAcronyms(g.mods)
                r.scoringType = g.scoringType
                r.teamType = g.teamType
                r.roundID = g.roundID.toInt()

                run {
                    val isVs = (g.teamType == "team-vs")

                    val total =
                        g.scores.map(MatchScore::score).filter { it > 10000 }.stream().reduce(0) { a, b -> a + b }
                            .toLong()

                    val red = if (isVs) {
                        g.scores.stream().filter { it.playerStat.team == "red" }.map(MatchScore::score)
                            .filter { it > 10000 }.reduce(0) { a, b -> a + b }.toLong()
                    } else {
                        0L
                    }

                    val blue = if (isVs) {
                        g.scores.filter { it.playerStat.team == "blue" }.map(MatchScore::score).filter { it > 10000 }
                            .stream().reduce(0) { a, b -> a + b }.toLong()
                    } else {
                        0L
                    }

                    val isRedWin = isVs && red > blue
                    val isBlueWin = isVs && red < blue

                    r.teamScore = LegacyMatch.MatchRound.TeamScore(total, red, blue)
                    r.winningTeam = if (isBlueWin) "blue"
                    else if (isRedWin) "red" else if (isVs) null else "none"
                    r.winningTeamScore = if (isBlueWin) blue
                    else if (isRedWin) red
                    else if (isVs) blue
                    else {
                        (g.scores.maxOfOrNull(MatchScore::score) ?: 0).toLong()
                    }
                }
            }
            e.userID = me.userID
            e.eventID = me.eventID
            e.round = r
            e.timestamp = me.timestamp

            e.detail = LegacyMatch.MatchEvent.Detail(me.detail.type, me.detail.text)

            es.add(e)
        }

        match.events = es

        return match
    }

     */

    operator fun plusAssign(match: Match) {
        // 更新玩家
        if (match.players.isNotEmpty()) {
            val userSet = players.map { it.userID }.toSet()
            val newUsers = match.players.filter { it.userID in userSet }
            players.addAll(newUsers)
        }

        // 更新状态
        statistics = match.statistics
        latestEventID = match.latestEventID
        firstEventID = match.firstEventID

        if (match.events.isEmpty()) return

        // 处理空 score 对局
        val lastGame = this.events.lastOrNull {it.round != null}

        if (lastGame?.round?.scores?.isEmpty() == true) {
            val replacer = match.events.lastOrNull { r -> lastGame.eventID == r.eventID }
            if (replacer != null) {
                val index = this.events.indexOf(lastGame)
                this.events[index] = replacer
            }
        }

        when {
            // 插入新事件
            events.last().eventID < match.events.first().eventID -> {
                events.addAll(match.events)
            } // 插入旧事件
            events.first().eventID > match.events.last().eventID -> {
                events.addAll(0, match.events)
            } // 中间插入
            events.last().eventID < match.events.last().eventID -> {
                events.removeIf { it.eventID >= match.events.first().eventID }
                events.addAll(match.events)
            } // 中间插入
            events.first().eventID > match.events.first().eventID -> {
                events.removeIf { it.eventID <= match.events.last().eventID }
                events.addAll(0, match.events)
            }
        }
    }

    companion object {
        fun Match.append(m: Match) {
            if (m.events.isEmpty()) return
            // 合并事件

            if (this.events.first().eventID > m.events.last().eventID) {
                // 新事件在前
                this.events.addAll(0, m.events)
            } else if (events.last().eventID < m.events.first().eventID) {
                // 新事件在后
                this.events.addAll(m.events)
            } else if (events.first().eventID > m.events.first().eventID) {
                // 在中间
                events.removeIf { e: MatchEvent -> e.eventID <= m.events.last().eventID }
                this.events.addAll(0, m.events)
            } else if (events.last().eventID < m.events.last().eventID) {
                // 在中间
                events.removeIf { e: MatchEvent -> e.eventID >= m.events.first().eventID }
                this.events.addAll(m.events)
            }

            if (this.players.isEmpty()) {
                this.players.addAll(m.players)
            } else {
                val p = m.players.subtract(this.players.toSet())

                this.players.addAll(p)
            }

            //更新状态
            this.statistics = m.statistics
            this.latestEventID = m.latestEventID
            this.firstEventID = m.firstEventID
        }
    }
}
